drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1670 @@
+end
+
+-----------------------------------------------------
+-- Number Of Year Day Schedules Supported Per User --
+-----------------------------------------------------
+local function max_year_schedule_of_user_handler(driver, device, ib, response)
+  device:emit_event(capabilities.lockSchedules.yearDaySchedulesPerUser(ib.data.value, {visibility = {displayed = false}}))
+end
+
+-- Capability Handler
+-----------------
+-- Lock/Unlock --
+-----------------
+local function handle_lock(driver, device, command)
+  local ep = device:component_to_endpoint(command.component)
+  local cota_cred = device:get_field(lock_utils.COTA_CRED)
+  if cota_cred then
+    device:send(
+      DoorLock.server.commands.LockDoor(device, ep, cota_cred)
+    )
+  else
+    device:send(DoorLock.server.commands.LockDoor(device, ep))
+  end
+end
+
+local function handle_unlock(driver, device, command)
+  local ep = device:component_to_endpoint(command.component)
+  local cota_cred = device:get_field(lock_utils.COTA_CRED)
+  if cota_cred then
+    device:send(
+      DoorLock.server.commands.UnlockDoor(device, ep, cota_cred)
+    )
+  else
+    device:send(DoorLock.server.commands.UnlockDoor(device, ep))
+  end
+end
+
+----------------
+-- User Table --
+----------------
+local function add_user_to_table(device, userIdx, usrType)
+  -- Get latest user table
+  local user_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME,
+    {}
+  ))
+
+  -- Add new entry to table
+  table.insert(user_table, {userIndex = userIdx, userType = usrType})
+  device:emit_event(capabilities.lockUsers.users(user_table, {visibility = {displayed = false}}))
+end
+
+local function update_user_in_table(device, userIdx, usrType)
+  -- Get latest user table
+  local user_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME,
+    {}
+  ))
+
+  -- Find user entry
+  local i = 0
+  for index, entry in pairs(user_table) do
+    if entry.userIndex == userIdx then
+      i = index
+      break
+    end
+  end
+
+  -- Update user entry
+  if i ~= 0 then
+    user_table[i].userType = usrType
+    device:emit_event(capabilities.lockUsers.users(user_table, {visibility = {displayed = false}}))
+  end
+end
+
+local function delete_user_from_table(device, userIdx)
+  -- If User Index is ALL_INDEX, remove all entry from the table
+  if userIdx == ALL_INDEX then
+    device:emit_event(capabilities.lockUsers.users({}, {visibility = {displayed = false}}))
+    return
+  end
+
+  -- Get latest user table
+  local user_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME,
+    {}
+  ))
+
+  -- Re-create user table
+  for index, entry in pairs(user_table) do
+    if entry.userIndex == userIdx then
+      table.remove(user_table, index)
+      break
```

Comment:
Is it possible that a userIdx could be accessed multiple times? I guess, to be clear, should we be `break`ing here?

Reply:
I don't understand what you mean. Do you mean we should check all entries and remove the entries that has same user index?
we don't need to check all entries. the all user indexes are unique so if we find the entry with same user index, we don't need to check rest entries anymore.

Reply:
ah, perfect. you answered my question, thank you!